File: src/factory/PomFactory.java
package factory;

import com.intellij.psi.PsiElement;
import com.intellij.psi.xml.XmlElement;
import com.intellij.psi.xml.XmlTag;
import com.intellij.psi.xml.XmlText;
import com.intellij.psi.xml.XmlToken;
import domain.Dependency;
import domain.XmlDependency;
import domain.Pom;

import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

import static domain.constant.XmlNodeName.ARTIFACT_ID;
import static domain.constant.XmlNodeName.DEPENDENCY;
import static domain.constant.XmlNodeName.GROUP_ID;
import static domain.constant.XmlNodeName.PROJECT;
import static domain.constant.XmlNodeName.VERSION;

/**
 *
 */
public class PomFactory {
    private static final String VARIABLE = "^\\s*\\$\\s*\\{.+}\\s*$";

    /**
     *
     * @param currentPsiElement
     * @return
     */
    public Pom create(final PsiElement currentPsiElement) {
        final var currentXmlToken = castOrNull(currentPsiElement, XmlToken.class);
        Pom pom = null;
        if (currentPsiElement != null) {
            pom = create(currentXmlToken);
        }
        return pom;
    }

    private Pom create(final XmlToken currentXmlToken) {
        XmlText currentText = castOrNull(currentXmlToken.getParent(), XmlText.class);
        XmlTag currentTag = castOrNull(currentXmlToken.getParent(), XmlTag.class);
        if (currentText != null) {
            currentTag = castOrNull(currentText.getParent(), XmlTag.class);
        } else if (currentTag != null) {
            final var xmlTexts = extractXmlElement(currentTag, XmlText.class);
            if (xmlTexts.size() == 1) {
                currentText = xmlTexts.get(0);
            }
        }
        Pom pom = null;
        if (currentText != null && currentTag != null) {
            pom = createPom(currentText, currentTag);
        }
        return pom;
    }

    private Pom createPom(final XmlText version, final XmlTag versionTag) {
        final var noAttributesForVersionTag = versionTag.getAttributes().length == 0;
        final var isVersionTag = VERSION.getXmlName().equals(versionTag.getName());
        final var dependency = castOrNull(versionTag.getParent(), XmlTag.class);
        Pom pom = null;
        if (isVersionTag && noAttributesForVersionTag && dependency != null && !isVariable(version.getText())) {
            final var project = findBack(dependency, PROJECT.getXmlName());
            final var isDependencyTag = DEPENDENCY.getXmlName().equals(dependency.getName());
            final var noAttributesForDependencyTag = dependency.getAttributes().length == 0;
            if (project != null && isDependencyTag && noAttributesForDependencyTag) {
                pom = createPom(project, dependency, version);
            }
        }
        return pom;
    }

    private XmlTag findBack(final XmlTag xmlTag, final String name) {
        final var parent = castOrNull(xmlTag.getParent(), XmlTag.class);
        final XmlTag ret;
        if (parent != null) {
            if (name.equals(parent.getName())) {
                ret = parent;
            } else {
                ret = findBack(parent, name);
            }
        } else {
            ret = null;
        }
        return ret;
    }

    private Pom createPom(final XmlTag project, final XmlTag dependencyTag, final XmlText versionText) {
        return Pom.builder()
                .withNamespace(dependencyTag.getNamespace())
                .withProject(project)
                .withXmlDependency(XmlDependency.builder()
                        .withVersionXmlText(versionText)
                        .withDependency(createDependency(dependencyTag, versionText.getText()))
                        .build())
                .build();
    }

    private Dependency createDependency(final XmlTag dependencyTag, final String version) {
        final var xmlTagChildren = extractXmlElement(dependencyTag, XmlTag.class);
        final var groupIdTag = extractXmlTagByName(xmlTagChildren, GROUP_ID.getXmlName());
        final var artifactIdTag = extractXmlTagByName(xmlTagChildren, ARTIFACT_ID.getXmlName());
        Dependency dependency = null;
        if (groupIdTag != null && artifactIdTag != null) {
            dependency = createDependency(version, groupIdTag, artifactIdTag);
        }
        return dependency;
    }

    private <T extends XmlElement> List<T> extractXmlElement(final XmlTag dependencyTag, final Class<T> clazz) {
        return Arrays.stream(dependencyTag.getValue().getChildren())
                .filter(clazz::isInstance)
                .map(clazz::cast)
                .collect(Collectors.toList());
    }

    private Dependency createDependency(final String version, final XmlTag groupIdTag, final XmlTag artifactIdTag) {
        final var groupId = groupIdTag.getValue().getText();
        final var artifactId = artifactIdTag.getValue().getText();

        return Dependency.builder()
                .withArtifactId(artifactId)
                .withGroupId(groupId)
                .withVersion(version)
                .build();
    }

    private boolean isVariable(final String version) {
        return version.trim().matches(VARIABLE);
    }

    private XmlTag extractXmlTagByName(final List<XmlTag> xmlTagChildren, final String name) {
        return xmlTagChildren.stream()
                .filter(xmlTag -> name.equals(xmlTag.getName()))
                .findFirst()
                .orElse(null);
    }

    private <T, S extends T> S castOrNull(final T el, final Class<S> clazz) {
        S ret = null;
        if (el != null && clazz != null && clazz.isAssignableFrom(el.getClass())) {
            ret = clazz.cast(el);
        }
        return ret;
    }
}

File: src/factory/PomFactory.kt
package factory

import com.intellij.psi.PsiElement
import com.intellij.psi.xml.*
import domain.Dependency
import domain.Pom
import domain.XmlDependency
import domain.constant.XmlNodeName
import java.util.*
import java.util.stream.Collectors

class PomFactory {
    companion object {
        private const val VARIABLE = "^\\s*\\$\\s*\\{.+}\\s*$"
    }

    fun create(currentPsiElement: PsiElement): Pom? {
        val currentXmlToken = castOrNull(currentPsiElement, XmlToken::class.java)
        return if (currentXmlToken != null) {
            create(currentXmlToken)
        } else {
            null
        }
    }

    private fun create(currentXmlToken: XmlToken): Pom? {
        var currentText = castOrNull(currentXmlToken.parent, XmlText::class.java)
        var currentTag = castOrNull(currentXmlToken.parent, XmlTag::class.java)
        if (currentText != null) {
            currentTag = castOrNull(currentText.parent, XmlTag::class.java)
        } else if (currentTag != null) {
            val xmlTexts = extractXmlElement(currentTag, XmlText::class.java)
            if (xmlTexts.size == 1) {
                currentText = xmlTexts[0]
            }
        }
        return if (currentText != null && currentTag != null) {
            createPom(currentText, currentTag)
        } else {
            null
        }
    }

    private fun createPom(version: XmlText, versionTag: XmlTag): Pom? {
        val noAttributesForVersionTag = versionTag.attributes.isEmpty()
        val isVersionTag = XmlNodeName.VERSION.xmlName == versionTag.name
        val dependency = castOrNull(versionTag.parent, XmlTag::class.java)
        var pom: Pom? = null
        if (isVersionTag && noAttributesForVersionTag && dependency != null && !isVariable(version.text)) {
            val project = findBack(dependency, XmlNodeName.PROJECT.xmlName)
            val isDependencyTag = XmlNodeName.DEPENDENCY.xmlName == dependency.name
            val noAttributesForDependencyTag = dependency.attributes.isEmpty()
            if (project != null && isDependencyTag && noAttributesForDependencyTag) {
                pom = createPom(project, dependency, version)
            }
        }
        return pom
    }

    private fun findBack(xmlTag: XmlTag, name: String): XmlTag? =
            castOrNull(xmlTag.parent, XmlTag::class.java)?.let {
                if (name == it.name) {
                    it
                } else {
                    findBack(it, name)
                }
            }

    private fun createPom(project: XmlTag, dependencyTag: XmlTag, versionText: XmlText) =
            createDependency(dependencyTag, versionText.text)?.let {
                val xmlDependency = XmlDependency(it, versionText)
                Pom(dependencyTag.namespace, project, xmlDependency)
            }

    private fun createDependency(dependencyTag: XmlTag, version: String): Dependency? {
        val xmlTagChildren = extractXmlElement(dependencyTag, XmlTag::class.java)
        val groupIdTag = extractXmlTagByName(xmlTagChildren, XmlNodeName.GROUP_ID.xmlName)
        val artifactIdTag = extractXmlTagByName(xmlTagChildren, XmlNodeName.ARTIFACT_ID.xmlName)
        return if (groupIdTag != null && artifactIdTag != null) {
            createDependency(version, groupIdTag, artifactIdTag)
        } else {
            null
        }
    }

    private fun <T : XmlElement?> extractXmlElement(dependencyTag: XmlTag, clazz: Class<T>): List<T> {
        return Arrays.stream(dependencyTag.value.children)
                .filter { obj: XmlTagChild? -> clazz.isInstance(obj) }
                .map { obj: XmlTagChild? -> clazz.cast(obj) }
                .collect(Collectors.toList())
    }

    private fun createDependency(version: String, groupIdTag: XmlTag, artifactIdTag: XmlTag): Dependency {
        val groupId = groupIdTag.value.text
        val artifactId = artifactIdTag.value.text
        return Dependency(
                groupId,
                artifactId,
                version
        )
    }

    private fun isVariable(version: String) =
            version.matches(Regex(VARIABLE))

    private fun extractXmlTagByName(xmlTagChildren: List<XmlTag>, name: String) =
            xmlTagChildren.firstOrNull { name == it.name }

    private fun <T, S : T?> castOrNull(el: T?, clazz: Class<S>?) =
            if (el != null && clazz != null && clazz.isAssignableFrom(el.javaClass)) {
                clazz.cast(el)
            } else {
                null
            }
}
